src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -4026,6 +4026,417 @@ class StubGenerator: public StubCodeGenerator {
     return start;
   }
 
+  // ChaCha20 block function.  This version parallelizes 4 quarter
+  // round operations at a time.  It uses 16 SIMD registers to
+  // produce 4 blocks of key stream.
+  //
+  // state (int[16]) = c_rarg0
+  // keystream (byte[256]) = c_rarg1
+  // return - number of bytes of keystream (always 256)
+  address generate_chacha20Block_qr() {
+    // Macro to expand the quarter round for the a/b/c/d vectors
+    // for a given state block.
+    // -------------------------------------------------------
+    // a += b, d ^= a, d <<<= 16
+    // c += d, b ^= c, b <<<= 12
+    // a += b, d ^= a, d <<<= 8
+    // c += d, b ^= c, b <<<= 7
+    #define FOUR_QTR_ROUND(a, b, c, d)      \
+        __ addv(a, __ T4S, a, b);           \
+        __ eor(d, __ T16B, d, a);           \
+        __ rev32(d, __ T8H, d);             \
+                                            \
+        __ addv(c, __ T4S, c, d);           \
+        __ eor(scratch, __ T16B, b, c);     \
+        __ ushr(b, __ T4S, scratch, 20);    \
+        __ sli(b, __ T4S, scratch, 12);     \
+                                            \
+        __ addv(a, __ T4S, a, b);           \
+        __ eor(d, __ T16B, d, a);           \
+        __ tbl(d, __ T16B, d,  1, lrot8Tbl);\
+                                            \
+        __ addv(c, __ T4S, c, d);           \
+        __ eor(scratch, __ T16B, b, c);     \
+        __ ushr(b, __ T4S, scratch, 25);    \
+        __ sli(b, __ T4S, scratch, 7);      \
+
+    // Lane shifting used for column-to-diagonal and vice-versa
+    #define SHIFT_LANES(b, bCnt, c, cCnt, d, dCnt)  \
+        __ ext(b, __ T16B, b, b, bCnt);             \
+        __ ext(c, __ T16B, c, c, cCnt);             \
+        __ ext(d, __ T16B, d, d, dCnt);             \
+
```

Comment:
There's a fairly extensive use of macros here for the rounds, but I don't think there's any need for them to be macros. `SHIFT_LANES` and all the other macros here should be functions. This would reduce the size of the libjvm.so binary.

Reply:
Thanks for the feedback.  I've been wondering if I might need something like a macroAssembler_<arch>_chapoly.cpp file to handle these kinds of things and future functions for Poly1305 when I start in on that.  I wasn't aware of the impact on libjvm.so going the macro approach versus functions.  I'll pull these out to functions.